Linker errors in tests on MacOS

Failing tests:
- lac/constraints_local_to_global_rect
```
1894: Undefined symbols for architecture x86_64:
1894:   "void dealii::ConstraintMatrix::distribute_local_to_global<dealii::FullMatrix<double> >(dealii::FullMatrix<dealii::FullMatrix<double>::value_type> const&, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, dealii::ConstraintMatrix const&, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, dealii::FullMatrix<double>&) const", referenced from:
1894:       test() in constraints_local_to_global_rect.cc.o
```
- matrix_free/assemble_matrix_01
```
2394: Undefined symbols for architecture x86_64:
2394:   "void dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::reinit<2>(dealii::Quadrature<1> const&, dealii::FiniteElement<2, 2> const&, unsigned int)", referenced from:
2394:       dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::ShapeInfo<2>(dealii::Quadrature<1> const&, dealii::FiniteElement<2, 2> const&, unsigned int) in assemble_matrix_01.cc.o
2394:   "void dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::reinit<3>(dealii::Quadrature<1> const&, dealii::FiniteElement<3, 3> const&, unsigned int)", referenced from:
2394:       dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::ShapeInfo<3>(dealii::Quadrature<1> const&, dealii::FiniteElement<3, 3> const&, unsigned int) in assemble_matrix_01.cc.o
```
- matrix_free/assemble_matrix_02: Same as above
- matrix_free/assemble_matrix_03 : Same as above
- matrix_free/matrix_vector_15: Same as above
- matrix_free/matrix_vector_16: Same as above

--- a/doc/news/changes/minor/20170809ArndtMaierPelteret
+++ b/doc/news/changes/minor/20170809ArndtMaierPelteret
@@ -1,0 +1,4 @@
+Fixed: Debugged and fixed a long standing linker error due to missing
+instantiations of ShapeInfo::reinit variants.
+<br>
+(Jean-Paul Pelteret, Matthias Maier, Daniel Arndt, 2017/08/09)

--- a/source/matrix_free/matrix_free.inst.in
+++ b/source/matrix_free/matrix_free.inst.in
@@ -39,6 +39,12 @@ for (deal_II_dimension : DIMENSIONS)
     template void internal::MatrixFreeFunctions::ShapeInfo<float>::reinit
     <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
                         <deal_II_dimension,deal_II_dimension> &, const unsigned int);
+    template void internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<double> >::reinit
+    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
+                        <deal_II_dimension,deal_II_dimension> &, const unsigned int);
+    template void internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<float > >::reinit
+    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
+                        <deal_II_dimension,deal_II_dimension> &, const unsigned int);
 #endif
 }
 

--- a/source/matrix_free/matrix_free.inst.in
+++ b/source/matrix_free/matrix_free.inst.in
@@ -33,18 +33,33 @@ for (deal_II_dimension : DIMENSIONS)
     template struct internal::MatrixFreeFunctions::MappingInfo<deal_II_dimension,float>;
 
 #ifndef DEAL_II_MSVC
-    template void internal::MatrixFreeFunctions::ShapeInfo<double>::reinit
-    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
-                        <deal_II_dimension,deal_II_dimension> &, const unsigned int);
-    template void internal::MatrixFreeFunctions::ShapeInfo<float>::reinit
-    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
-                        <deal_II_dimension,deal_II_dimension> &, const unsigned int);
-    template void internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<double> >::reinit
-    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
-                        <deal_II_dimension,deal_II_dimension> &, const unsigned int);
-    template void internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<float > >::reinit
-    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
-                        <deal_II_dimension,deal_II_dimension> &, const unsigned int);
+    template
+    void
+    internal::MatrixFreeFunctions::ShapeInfo<double>::
+    reinit<deal_II_dimension>(const Quadrature<1> &,
+                              const FiniteElement<deal_II_dimension,deal_II_dimension> &,
+                              const unsigned int);
+
+    template
+    void
+    internal::MatrixFreeFunctions::ShapeInfo<float>::
+    reinit<deal_II_dimension>(const Quadrature<1> &,
+                              const FiniteElement<deal_II_dimension,deal_II_dimension> &,
+                              const unsigned int);
+
+    template
+    void
+    internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<double> >::
+    reinit<deal_II_dimension>(const Quadrature<1> &,
+                              const FiniteElement<deal_II_dimension,deal_II_dimension> &,
+                              const unsigned int);
+
+    template
+    void
+    internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<float > >::
+    reinit<deal_II_dimension>(const Quadrature<1> &,
+                              const FiniteElement<deal_II_dimension,deal_II_dimension> &,
+                              const unsigned int);
 #endif
 }
 

